### PR TITLE
React coa

### PR DESCRIPTION
**Pull request by charliehuang09**
- Implement React into langchain
- Used this tutorial/code: https://js.langchain.com/docs/modules/agents/how_to/custom_llm_agent

### CODE REVIEW

File: scripts/check.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+pnpm lint:fix
+pnpm i build
```
Comment (maxstrid) [inline scripts/check.sh:2]: This is whats adding build as a dependency, you only need to do pnpm i.

File: src/lib/tools.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { SerpAPI } from 'langchain/tools';
+import { Calculator } from 'langchain/tools/calculator';
+import { DynamicTool, DynamicStructuredTool } from 'langchain/tools';
+import { QdrantClient } from '@qdrant/js-client-rest';
+import { OpenAIEmbeddings } from 'langchain/embeddings/openai';
+import { z } from 'zod';
+
+function getTools() {
+	if (import.meta.env.VITE_SERPAPI_API_KEY == undefined) {
```
Comment (maxstrid) [inline src/lib/tools.ts:9]: This is good to do this, I actually think we should do this with all of our API keys. We might want to have some code which does this in server.ts maybe as a separate PR.

File: src/lib/tools.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		throw console.warn('SERPAPI_API_KEY is undefined');
+	}
+	const tools = [
+		new DynamicTool({
+			//for testing remove when merge
+			name: 'FRC971Database',
+			description: 'useful to get information about FRC971. Input should be a search query.',
+			func: async (query: string) => {
+				const embeddings = new OpenAIEmbeddings({ openAIApiKey: process.env.OPENAI_API_KEY });
+				const client = new QdrantClient({ host: 'localhost', port: 6333 });
```
Comment (maxstrid) [inline src/lib/tools.ts:20]: I'd prefer for ChatCompletion to hold a QdrantClient and for it to pass it into the getTools function so we don't create a new one every time we get our tools.

File: src/lib/tools.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+				return output;
+			}
+		}),
+		new SerpAPI(import.meta.env.VITE_SERPAPI_API_KEY, {
+			location: 'Austin,Texas,United States',
+			hl: 'en',
+			gl: 'us'
+		}),
+		new DynamicStructuredTool({
+			//for testing remove when merge
```
Comment (maxstrid) [inline src/lib/tools.ts:36]: Could you expand on this? I'm not really sure why we need this tool

File: src/routes/+page.svelte [hunk 5]
```diff
@@ -96,15 +78,9 @@
 			placeholder="Enter your message..."
 			class="m-auto outline-none active:border-none rounded-md p-2 flex-1"
 		/>
-		{#if $completionState == CompletionState.Loading}
-			<button on:click={stop}>
-				<Icon src={Stop} class="w-6 h-6 text-gray-500 hover:text-gray-950" />
-			</button>
-		{:else}
-			<button on:click={send}>
```
Comment (maxstrid) [inline src/routes/+page.svelte:104]: We should still be streaming here. If we want to remove it from this app I would keep it in the markdown pr because thats the reason for it. But we can stream agents in langchain as far as I understand, they let you stream only the final message.

File: src/lib/completion.ts [hunk 6]
```diff
@@ -1,131 +1,144 @@
...
-	private qdrant_client: QdrantClient;
-	private qdrant_collection: string;
+	private executor: AgentExecutor | undefined;
+	private openai_api_key: string;
+	private model_name;
+	private verbose: boolean;
 
 	constructor(
 		openai_api_key: string,
+		verbose: boolean,
```
Comment (maxstrid) [inline src/lib/completion.ts:32]: I'd prefer for this to be in the optional config object, just to be more clear what it is when calling.

File: src/lib/completion.ts [hunk 7]
```diff
@@ -1,131 +1,144 @@
...
 		{
-			openai_model = DEFAULT_MODEL,
-			qdrant_collection = DEFAULT_COLLECTION
```
Comment (maxstrid) [inline src/lib/completion.ts:26]: This is another reason to keep the qdrant collection in this class because I want someone developing locally to be able to use whatever collection they want as this allows multiple iterations on the database.

File: src/lib/completion.ts [hunk 8]
```diff
@@ -1,131 +1,144 @@
...
+		this.verbose = verbose;
+		this.model_name = openai_model;
+		this.openai_api_key = openai_api_key;
 
 		this.embeddings_model = new OpenAIEmbeddings({
 			openAIApiKey: openai_api_key,
 			modelName: 'text-embedding-ada-002'
 		});
-		this.qdrant_client = new QdrantClient({
-			url: 'http://' + (process.env.QDRANT_HOST ?? 'localhost') + ':6333'
```
Comment (maxstrid) [inline src/lib/completion.ts:46]: We also need this so that it works with docker, as with docker compose the url will be qdrantdb

File: src/lib/completion.ts [hunk 9]
```diff
@@ -1,131 +1,144 @@
...
+
+		const formatted = [
+			PREFIX,
+			toolString,
+			TOOL_INSTRUCTIONS_TEMPLATE.replace('{tool_names}', toolNames),
+			SUFFIX.replace('{input}', values.input),
+			agentScratchpad
+		].join('');
+		if (this.verbose) {
+			console.log(colors.fg.green, 'Model prompt: ', colors.style.reset);
```
Comment (maxstrid) [inline src/lib/completion.ts:101]: nit: We should probably use a proper logging library but if you don't have the bandwidth for this it isn't needing for release so you could ignore this.

File: src/lib/completion.ts [hunk 10]
```diff
@@ -1,131 +1,144 @@
...
+			const parts = content.split('Final Answer:');
+			const input = parts[parts.length - 1].trim();
+			const finalAnswers = { output: input };
+			return { log: content, returnValues: finalAnswers };
+		}
+		const match = /Action: (.*)\nAction Input: (.*)/s.exec(content);
+		if (match == null) {
+			console.warn('Could not parse output');
+			console.warn(text);
+			process.exit(42);
```
Comment (maxstrid) [inline src/lib/completion.ts:123]: We should't ever exit from here, we need to make sure the backend stays up. Even if we can't continue with the completion I want us to just propagate our errors and give them to the user so they could submit a bug report.

File: vite.config.ts [hunk 11]
```diff
@@ -1,5 +1,8 @@
 import { sveltekit } from '@sveltejs/kit/vite';
 import { defineConfig } from 'vite';
+import dotenv from 'dotenv';
+
+dotenv.config();
```
Comment (maxstrid) [inline vite.config.ts:5]: Why are we using dotenv here?

File: src/routes/api/completion/+server.ts [hunk 12]
```diff
@@ -3,19 +3,19 @@ import { ChatbotCompletion } from '$lib';
...
 
-	const completion_manager = new ChatbotCompletion(import.meta.env.VITE_OPENAI_API_KEY!, {
+	const completion_manager = new ChatbotCompletion(import.meta.env.VITE_OPENAI_API_KEY!, true, {
 		openai_model: import.meta.env.VITE_OPENAI_MODEL_NAME!
 	});
+	// ChatbotCompletion(openaikey, verbose, {model_name})
+	await completion_manager.setup();
 
-	const stream = await completion_manager.query(chatHistory);
+	const output = await completion_manager.query(chatHistory, input);
```
Comment (maxstrid) [inline src/routes/api/completion/+server.ts:14]: We need to handle the error returned by query.

File: src/lib/completion.ts [hunk 13]
```diff
@@ -1,131 +1,144 @@
...
-		const qdrant_results = await this.qdrant_client.search(this.qdrant_collection, {
-			vector: query_embedding,
-			limit: k,
-			with_vector: true
+	public async setup() {
+		const tools = getTools();
+		const model = new ChatOpenAI({
+			openAIApiKey: this.openai_api_key,
+			modelName: this.model_name,
+			temperature: 0,
```
Comment (maxstrid) [inline src/lib/completion.ts:55]: What's the reason for temp 0 here?

**Review by maxstrid — CHANGES_REQUESTED** [review]

File: src/lib/completion.ts (hunk 10 above)
Comment (maxstrid) [reply src/lib/completion.ts:123]: You don't need to show errors in any special way to the users, I just want to make sure its returned to the frontend.

File: src/lib/completion.ts (hunk 13 above)
Comment (charliehuang09) [reply src/lib/completion.ts:55]: To make the model as deterministic as possible

File: src/lib/tools.ts (hunk 3 above)
Comment (charliehuang09) [reply src/lib/tools.ts:20]: It would be hard to pass in the qdrant client into the function because I would need to create a custom agent execute tool function. However, we can set embeddings and client as global variables. What do you think?

File: src/routes/api/completion/+server.ts (hunk 12 above)
Comment (charliehuang09) [reply src/routes/api/completion/+server.ts:14]: I am working on making the error less common by making the llm listen to instructions.

File: src/lib/tools.ts (hunk 3 above)
Comment (charliehuang09) [reply src/lib/tools.ts:19]: Nvm, I think I found a solution

File: vite.config.ts (hunk 11 above)
Comment (charliehuang09) [reply vite.config.ts:5]: I was testing and accidentily left this in.

File: src/routes/+page.svelte (hunk 5 above)
Comment (charliehuang09) [reply src/routes/+page.svelte:104]: Why do we want to use stream?

File: src/lib/completion.ts (hunk 10 above)
Comment (charliehuang09) [reply src/lib/completion.ts:144]: What should I return then?

File: src/lib/completion.ts (hunk 9 above)
Comment (charliehuang09) [reply src/lib/completion.ts:122]: What do you mean by a proper logging library?

File: src/routes/+page.svelte (hunk 5 above)
Comment (maxstrid) [reply src/routes/+page.svelte:104]: Streaming is a lot better on the user side, it lets you see if you like what its completing and decide if you want to stop and send a different message. It just makes the user experience overall better.

File: src/lib/completion.ts (hunk 9 above)
Comment (maxstrid) [reply src/lib/completion.ts:122]: Some library that supports things like logging to a file, thats my main thing. I want the ability to have file logs to look at later.

File: src/lib/completion.ts (hunk 10 above)
Comment (maxstrid) [reply src/lib/completion.ts]: I would throw an error when this happens and catch it up in the stack.

File: src/routes/api/completion/+server.ts (hunk 12 above)
Comment (maxstrid) [reply src/routes/api/completion/+server.ts:15]: Even if the error is less common we still need to make sure that the error is handled properly

File: src/lib/completion.ts (hunk 9 above)
Comment (charliehuang09) [reply src/lib/completion.ts:122]: Oh, ok I will do it after I finish fine tuning because it is not very urgent right now

Comment (charliehuang09) [reply src/lib/completion.ts:126]: Actually, I will do logging right now because it is useful to generate data for finetuning.

File: src/lib/completion.ts (hunk 8 above)
Comment (charliehuang09) [reply src/lib/completion.ts:46]: This can be in a seperate PR

File: src/lib/completion.ts (hunk 10 above)
Comment (charliehuang09) [reply src/lib/completion.ts]: It should be fixed now

File: src/routes/api/completion/+server.ts (hunk 12 above)
Comment (charliehuang09) [reply src/routes/api/completion/+server.ts:17]: I will just output the unparsed result

File: scripts/fine_tune.py [hunk 14]
```diff
@@ -0,0 +1,12 @@
+from openai import OpenAI
+
+client = OpenAI()
+
+model_name = ""
```
Comment (maxstrid) [inline scripts/fine_tune.py:5]: This should be put inside of a main function

File: scripts/generate_fine_tune_data.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
+# bad name
+
+import os
+import json
+
+PATH = '../data/fine_tune'
+
+data = []
```
Comment (maxstrid) [inline scripts/generate_fine_tune_data.py:8]: We should also put this inside of a main function

File: src/lib/completion.ts [hunk 16]
```diff
@@ -1,131 +1,216 @@
...
+	private model_name;
+	private verbose: boolean;
+	private qdrantClient: QdrantClient;
+	private collection_name: string;
+	private history: ChatHistory[];
+	private do_history: boolean;
+	private generate_data: boolean;
+	private chain: string[];
+	private summaryBot: OpenAI;
+	private do_summaryBot: boolean;
```
Comment (maxstrid) [inline src/lib/completion.ts:38]: nit: I'd prefer for this to be use_summarybot and not do_summarybot

File: src/lib/completion.ts [hunk 17]
```diff
@@ -1,131 +1,216 @@
...
-	private model: ChatOpenAI;
 	private embeddings_model: OpenAIEmbeddings;
-
-	private qdrant_client: QdrantClient;
-	private qdrant_collection: string;
+	private executor: AgentExecutor | undefined;
+	private openai_api_key: string;
+	private model_name;
+	private verbose: boolean;
+	private qdrantClient: QdrantClient;
```
Comment (maxstrid) [inline src/lib/completion.ts:31]: nit: use snake case here for consistency

File: src/lib/completion.ts [hunk 18]
```diff
@@ -1,131 +1,216 @@
...
+	private collection_name: string;
+	private history: ChatHistory[];
+	private do_history: boolean;
+	private generate_data: boolean;
+	private chain: string[];
+	private summaryBot: OpenAI;
```
Comment (maxstrid) [inline src/lib/completion.ts:37]: nit: use snake case here for consistency

File: src/lib/completion.ts [hunk 19]
```diff
@@ -1,131 +1,216 @@
...
+		console.log(colors.fg.cyan, 'Model name: ' + this.model_name, colors.style.reset);
+		console.log('\n');
+		const tools = getTools(
+			this.qdrantClient,
+			this.collection_name,
+			this.embeddings_model,
+			this.summaryBot,
+			this.do_summaryBot
+		);
+		const model = new ChatOpenAI({
```
Comment (maxstrid) [inline src/lib/completion.ts:94]: creating the chatbot isn't async and shouldn't need to be called in this function. We can do it in the constructor.

File: src/lib/completion.ts [hunk 20]
```diff
@@ -1,131 +1,216 @@
...
+				})
+				.join('\n') + '\n';
+		const tools = getTools(
+			this.qdrantClient,
+			this.collection_name,
+			this.embeddings_model,
+			this.summaryBot,
+			this.do_summaryBot
+		); //to do seperate function for this
+		const intermediateSteps = values.intermediate_steps
```
Comment (maxstrid) [inline src/lib/completion.ts:134]: nit: make these variables snake case for consistency

File: src/lib/completion.ts [hunk 21]
```diff
@@ -1,131 +1,216 @@
...
-		});
-	}
-
-	private generate_history(history: ChatHistory[]): BaseMessage[] {
-		return history.map((message: { content: string; type: ChatHistoryType }) => {
-			if (message.type == ChatHistoryType.AI) {
-				return new AIMessage({ content: message.content });
-			} else {
-				return new HumanMessage({ content: message.content });
+	private formatMessages = async (values: InputValues) => {
```
Comment (maxstrid) [inline src/lib/completion.ts:119]: Add a comment here explaining what this function does or find ways to clean up the code we're doing because as it is this function is pretty confusing at a glance.

File: src/lib/completion.ts [hunk 22]
```diff
@@ -1,131 +1,216 @@
...
+
+		const formatted = [
+			system,
+			HISTORY,
+			history,
+			SUFFIX.replace('{input}', values.input),
+			agentScratchpad
+		].join('');
+		if (this.verbose) {
+			console.log(colors.fg.green, 'Model prompt: ', colors.style.reset);
```
Comment (maxstrid) [inline src/lib/completion.ts:157]: I wan't us to also be logging a timestamp here as well.

File: src/lib/completion.ts [hunk 23]
```diff
@@ -1,131 +1,216 @@
...
+	private executor: AgentExecutor | undefined;
+	private openai_api_key: string;
+	private model_name;
+	private verbose: boolean;
+	private qdrantClient: QdrantClient;
+	private collection_name: string;
+	private history: ChatHistory[];
+	private do_history: boolean;
+	private generate_data: boolean;
+	private chain: string[];
```
Comment (maxstrid) [inline src/lib/completion.ts:36]: Add a comment explaining what this is or rename it. It's pretty confusing when mixing it up with the langchain agents as well

File: src/lib/completion.ts [hunk 24]
```diff
@@ -1,131 +1,216 @@
...
-			...this.generate_history(history)
-		];
-
-		// TODO(max): Add history pruning based on token length
-
-		return await this.model.pipe(new BytesOutputParser()).stream(chat_history);
+		if (this.do_history) this.history = history;
+		const result = await this.executor.invoke({ input: input });
+		if (this.generate_data)
+			fs.writeFile('data/logs/' + input + '.jsonl', this.chain.join('\n'), (err) => {
```
Comment (maxstrid) [inline src/lib/completion.ts:211]: We should write the log even if we aren't querying in case we crash before we get to this step. This makes me think we need to be abstracting this logging logic into a separate commit or use a mature logging library.

If you decide to abstract the logging code make a new pr and remove the chain logic for now.

File: src/lib/completion.ts [hunk 25]
```diff
@@ -1,131 +1,216 @@
...
+				if (err) throw err;
```
Comment (maxstrid) [inline src/lib/completion.ts:212]: I don't like us throwing an error here. I have a feeling its good enough to put a console warn and move on as this isn't 100% necessary for continuing to provide output to the user.

File: src/lib/tools.ts [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+				const strResponse = response
+					.map((response) => response.payload!.pageContent as string)
+					.join('\n');
+				const prompt = SUMMARY.replace('{question}', query).replace('{text}', strResponse);
+				if (do_summaryBot) return await summaryBot.call(prompt);
+				return strResponse;
+			}
+		}),
+		new DynamicTool({
+			//for testing remove when merge
```
Comment (maxstrid) [inline src/lib/tools.ts:38]: Given this is close to being merged we should remove this tool

File: src/lib/tools.ts [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+					limit: 3
+				}); //strResponse bad name
+				const strResponse = response
+					.map((response) => response.payload!.pageContent as string)
+					.join('\n');
+				const prompt = SUMMARY.replace('{question}', query).replace('{text}', strResponse);
+				return await summaryBot.call(prompt);
+			}
+		}),
+		// new SerpAPI(import.meta.env.VITE_SERPAPI_API_KEY, {
```
Comment (maxstrid) [inline src/lib/tools.ts:55]: Remove this comment

File: src/lib/tools.ts [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+		// 	location: 'Austin,Texas,United States',
+		// 	hl: 'en',
+		// 	gl: 'us'
+		// }),
+		new Calculator()
```
Comment (maxstrid) [inline src/lib/tools.ts:60]: Is the calculator ever being used? If not remove it

File: src/lib/tools.ts [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+import { SUMMARY } from './prompt';
+
+function getTools(
+	qdrantClient: QdrantClient,
+	collection_name: string,
+	embeddings: OpenAIEmbeddings,
+	summaryBot: OpenAI,
+	do_summaryBot: boolean
+) {
+	if (import.meta.env.VITE_SERPAPI_API_KEY == undefined) {
```
Comment (maxstrid) [inline src/lib/tools.ts:15]: It looks like we don't need this api key. I think this can be removed.

File: src/routes/+page.svelte [hunk 30]
```diff
@@ -58,27 +53,14 @@
...
-			console.log(decoder.decode(chunk));
-			history[history.length - 1].content += decoder.decode(chunk);
-
-			// TODO(max): Find a way to lock the actual completion from OpenAI, right now this will still use the full tokens
-			if (cancelStream) {
-				cancelStream = false;
-				break;
-			}
-		}
+		history[history.length - 1].content = message;
```
Comment (maxstrid) [inline src/routes/+page.svelte:59]: We need to be doing streams here. I think it works pretty much the same as it does with normal langchain queries. But streams are really useful from a user perspective

**Review by maxstrid — CHANGES_REQUESTED** [review]

File: src/lib/completion.ts (hunk 24 above)
Comment (charliehuang09) [reply src/lib/completion.ts:211]: I think we should keep this logging because it is used to generate fine tuned data but we can use a actual logging libary to log for the devs.

File: src/lib/completion.ts (hunk 22 above)
Comment (charliehuang09) [reply src/lib/completion.ts:158]: Do you mean console.log(current_time)?

File: src/lib/tools.ts (hunk 28 above)
Comment (charliehuang09) [reply src/lib/tools.ts:60]: If the user asks a math question it will most likely be used but since the user probably will be asking questions about FIRST, I will delete it.

File: src/lib/tools.ts (hunk 4 above)
Comment (charliehuang09) [reply src/lib/tools.ts:38]: The reason that there are 2 seperate tools for searching is so that the model can be more specific. 
For example when the model wants to know how to get a FIRSTAward, it will use the FIRSTAWARDS tool. When it wants to get info on how our team works, it will use the frc971 tool.

File: src/routes/+page.svelte (hunk 30 above)
Comment (charliehuang09) [reply src/routes/+page.svelte:59]: There is a option to stream it but for react it only stream the agent steps. So the Final answer will still only be one chunk. Streaming could be useful for logging though

File: src/lib/tools.ts (hunk 26 above)
Comment (charliehuang09) [reply src/lib/tools.ts:38]: Which tool are you talking about?

File: src/lib/completion.ts (hunk 22 above)
Comment (maxstrid) [reply src/lib/completion.ts:164]: Yeah, just log the time along with the text

File: src/lib/tools.ts (hunk 26 above)
Comment (maxstrid) [reply src/lib/tools.ts:38]: The dynamicTool here, the comment says remove when merge

File: src/routes/+page.svelte (hunk 30 above)
Comment (maxstrid) [reply src/routes/+page.svelte:59]: Taking a look at this I think it by default streams only the final output and not the intermediate steps. See https://js.langchain.com/docs/modules/model_io/chat/streaming. But if that doesn't work we can make a callback handler and have it create a stream.

File: src/lib/tools.ts (hunk 4 above)
Comment (maxstrid) [reply src/lib/tools.ts:38]: You should change the scripts to match that then, because right now theres only 1 collection being made.

File: package.json [hunk 31]
```diff
@@ -39,7 +39,10 @@
 		"@steeze-ui/heroicons": "^2.2.3",
 		"@steeze-ui/icons": "^2.1.0",
 		"@steeze-ui/svelte-icon": "^1.5.0",
-		"langchain": "^0.0.198",
-		"tiktoken": "^1.0.11"
+		"build": "^0.1.4",
```
Comment (maxstrid) [inline package.json:42]: I don't think we using some of these dependencies. Can you clean some of this up? Ex: build and zod

File: check.sh [hunk 32]
```diff
@@ -0,0 +1,4 @@
+pnpm lint:fix
```
Comment (maxstrid) [inline check.sh:1]: I don't think this script needs to be committed. I'd prefer to leave this in the package.json

File: scripts/generate_fine_tune_data.py [hunk 33]
```diff
@@ -0,0 +1,51 @@
+# bad name
```
Comment (maxstrid) [inline scripts/generate_fine_tune_data.py:1]: This comment should probably be changed to clarify what you mean or to remove it fully

File: src/lib/completion.ts [hunk 34]
```diff
@@ -1,131 +1,223 @@
...
-
+	private executor: AgentExecutor | undefined;
+	private openai_api_key: string;
+	private model_name;
+	private verbose: boolean;
 	private qdrant_client: QdrantClient;
-	private qdrant_collection: string;
+	private collection_name: string;
+	private history: ChatHistory[];
+	private do_history: boolean;
```
Comment (maxstrid) [inline src/lib/completion.ts:34]: Could this name be changed to better clarify what it does. You should maybe also add a comment explaining what this means as well.

File: src/lib/completion.ts [hunk 35]
```diff
@@ -1,131 +1,223 @@
...
-		});
+			this.chain.push(JSON.stringify({ chatbot: content }));
+			const parts = content.split('Final Answer:');
+			const input = parts[parts.length - 1].trim();
+			const finalAnswers = { output: input };
+			return { log: content, returnValues: finalAnswers };
+		}
+		const match = /Action: (.*)\nAction Input: (.*)/s.exec(content);
+		if (match == null) {
+			// this.chain.slice(-1, 1);
```
Comment (maxstrid) [inline src/lib/completion.ts:186]: Remove this comment

File: src/lib/completion.ts [hunk 36]
```diff
@@ -1,131 +1,223 @@
...
+		this.executor = executor;
 	}
-
-	private generate_history(history: ChatHistory[]): BaseMessage[] {
-		return history.map((message: { content: string; type: ChatHistoryType }) => {
-			if (message.type == ChatHistoryType.AI) {
-				return new AIMessage({ content: message.content });
-			} else {
-				return new HumanMessage({ content: message.content });
+	private promptBuilder = async (values: InputValues) => {
```
Comment (maxstrid) [inline src/lib/completion.ts:120]: This should just be a function instead of a variable set to a closure for consistency

File: src/lib/tools.ts (hunk 26 above)
Comment (charliehuang09) [reply src/lib/tools.ts:34]: Oh, that was talking about something else that was supposed to be removed
